### PR TITLE
Add support for proxying containers

### PR DESCRIPTION
**Pull request by lucaslorentz**
Closes https://github.com/lucaslorentz/caddy-docker-proxy/issues/11

Changes:
* Add support for proxying directly to containers, without swarm services

@ibnesayeed Can you please take a look?

### CODE REVIEW

**Comment by lucaslorentz** [comment]
I wonder if we need to support proxying same caddy website to multiple containers.

Something like:
* Group containers by caddy.address.
* For each group:
  - generate caddy config using labels from first container
  - generate proxy directive targetting all containers IPs of that group

File: README.md [hunk 1]
```diff
@@ -1,27 +1,27 @@
...
 ## Introduction
-This plugin enables caddy to be used as a reverse proxy for Docker Swarm services.
+This plugin enables caddy to be used as a reverse proxy for Docker.
 
 ## How does it work?
-It scans Docker Swarm services metadata looking for labels indicating that the service should be exposed on caddy.
+It scans Docker metadata looking for labels indicating that the service or container should be exposed on caddy.
 
-Then it generates an in memory Caddyfile with website entries and proxies directives pointing to each Docker Service internal DNS name.
+Then it generates an in memory Caddyfile with website entries and proxies directives pointing to each Docker service dns name or container IP.
```
Comment (ibnesayeed) [inline README.md]: dns => DNS

File: README.md [hunk 2]
```diff
@@ -1,27 +1,27 @@
...
 
-Every time Docker services changes, it updates the Caddyfile and triggers a caddy zero-downtime reload.
+Every time Docker objects changes, it updates the Caddyfile and triggers a caddy zero-downtime reload.
```
Comment (ibnesayeed) [inline README.md]: Docker objects => a Docker object

**Comment by ibnesayeed** [comment]
> I wonder if we need to support proxying same caddy website to multiple containers.

That is not a bad idea. In case of Swarm mode, replicated services are already covered, but bare containers can also benefit from it. However, grouping by the dress is not be safe. We will have to use all three (or perhaps all) `caddy.*` labels combined for grouping.

**Comment by lucaslorentz** [comment]
Good.
I'm not sure if many users would have replicated bare containers on the same host.
We can tackle that in another PR/issue.

**Comment by pwFoo** [comment]
What's about docker compose services without swarm? Multiple backend container should be loadbalanced for the same public domain.

**Comment by ibnesayeed** [comment]
> @lucaslorentz: I'm not sure if many users would have replicated bare containers on the same host.

As per the following, it is very likely to have multiple replicas because Compose makes it easy to scale services and it's been there long before Swarm mode was added to the Docker core.

> @pwFoo: What's about docker compose services without swarm? Multiple backend container should be loadbalanced for the same public domain.

**Comment by lucaslorentz** [comment]
@pwFoo 

I think Docker Swarm Stacks is the new docker-compose.

Even for isolated machines, Docker Swarm would be beneficial because it ensures your containers are always running. As far as I know docker-compose only creates/destroy containers when you call the command, it doesn't monitors in order to ensure the desired number of replicas are running.

I would like to understand better your scenario in order to do the right features. Can you please describe it a bit further?

- Do you want to run caddy inside a container or outside?
- How many docker hosts caddy needs to interact with?
- Do you have multiple replicas for same application? Or are you talking about multiple applications sharing the same public domain under different paths?

**Comment by ibnesayeed** [comment]
> Even for isolated machines, Docker Swarm would be beneficial because it ensures your containers are always running.

While this is very true, I think the Swarm mode was a little late entry in the ecosystem and by that time many people were already using Compose. Docker is still actively maintaining Compose separately because the `docker stack` is not a complete replacement of the `docker-compose` command yet. There are many things desired from it if the core docker commands were to eliminate the need of compose, especially in the development mode. I think docker is working in that direction, but GitHub is full of various projects that come with the `docker-compose.yml` file and instructions on how to get started with compose.

That said, I think if we start grouping containers for proxying based on `${caddy.address}:${caddy.targetport}/${caddy.targetpath}` then it won't matter how they were created.

**Comment by lucaslorentz** [comment]
Makes sense, I will do a new PR.

**Comment by pwFoo** [comment]
Hello @lucaslorentz ,

will loadbalancing added soon?
At the moment If I increase replicas each request is routed to the first container / caddyfile entry.
```
http://192.168.199.22/httpd1 {
  proxy / 10.0.2.5:80
}
http://192.168.199.22/httpd1 {
  proxy / 10.0.2.6:80
}
```

It should be balanced to the backend containers:
```
http://192.168.199.22/httpd1 {
  proxy / 10.0.2.5:80 10.0.2.6:80
}
```

**Comment by lucaslorentz** [comment]
@pwFoo can you please create an issue with that?
Just as a reminder and we can track it there.

**Pull request merged**